src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Activity.cs

```diff
@@ -187,5 +187,42 @@ public async Task DeleteOldVersionBlobAsync([ActivityTrigger] IDurableActivityCo
         logger.LogInformation("Old blobs deleted successfully. Total size {TotalCount}", fileIdentifiers.Count);
     }
 
+    /// <summary>
+    /// Asynchronously delete the new blob when there is a failure while updating the study instances.
+    /// </summary>
+    /// <param name="context">Activity context which has list of watermarks to cleanup</param>
+    /// <param name="logger">A diagnostic logger.</param>
+    /// <returns>
+    /// A task representing the <see cref="CleanupNewVersionBlobAsync"/> operation.
+    /// </returns>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="context"/> or <paramref name="logger"/> is <see langword="null"/>.
+    /// </exception>
+    [FunctionName(nameof(CleanupNewVersionBlobAsync))]
+    public async Task CleanupNewVersionBlobAsync([ActivityTrigger] IDurableActivityContext context, ILogger logger)
+    {
+        EnsureArg.IsNotNull(context, nameof(context));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        IReadOnlyList<InstanceFileState> fileIdentifiers = context.GetInput<IReadOnlyList<InstanceFileState>>();
+
+        logger.LogInformation("Begin cleaning up new blobs. Total size {TotalCount}", fileIdentifiers.Count);
+
+        await Parallel.ForEachAsync(
+            fileIdentifiers.Where(f => f.NewVersion.HasValue),
+            new ParallelOptions
+            {
+                CancellationToken = default,
+                MaxDegreeOfParallelism = _options.MaxParallelThreads,
+            },
+            async (fileIdentifier, token) =>
+            {
+                await _updateInstanceService.DeleteInstanceBlobAsync(fileIdentifier.NewVersion.Value, token);
+            });
+
+        logger.LogInformation("New blobs deleted successfully. Total size {TotalCount}", fileIdentifiers.Count);
```

Comment:
Could you update to fileIdentifiers.Where(f => f.NewVersion.HasValue) count. Can you do the same for the previous activity